src/Microsoft.AspNetCore.Mvc.Core/ControllerAttribute.cs

```diff
@@ -0,0 +1,17 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore.Mvc
+{
+    /// <summary>
+    /// Indicates that the type and any derived types that this attribute is applied to
+    /// are considered a controller by the default controller discovery mechanism, unless
+    /// <see cref="NonControllerAttribute"/> is applied to any type in the hierarchy.
```

Comment:
@javiercn you're being more simplistic about the behaviour than users will expect. "Closest attribute wins" is fairly common and "this one always wins" is rather rare.